proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,167 @@
+        | try bt instr* (catch x instr*)+ end
+        | try bt instr* (catch x instr*)* catch_all instr* end
+        | try bt instr* delegate l
+        | try bt instr* unwind instr* end
+```
+
+### Modules
+
+Exceptions (definitions)
+
+```
+exn ::= export* exception exntype  | export* exception exntype import
+```
+
+Modules
+
+
+```
+mod ::= module ... exn*
+```
+
+## Validation (Typing)
+
+
+To verify that a `try...delegate l` instruction refers to a label surrounding the instructions of a try block (call this a try-label), introduce a `kind` attribute to labels in the validation context, which is set to `try` when the label is a try-label.
+
+Similarly, to verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we allow the `kind` attribute of labels in the validation context to be set to `catch` when the label is a catch-label. This addition is reflected in the execution rules, by the administrative instruction `caught` which introduces a label around the catching try-block.
+
+The original notation `label [t*]` is now a shortcut for `label {result [t*], kind ε}`.
+
+
+### Instructions
+
+
+```
+C_exn(x) = [t*] -> []
+--------------------------------
+C ⊢ throw x : [t1* t*] -> [t2*]
+
+
+C_label(l) =/= ε
+C_label(l).kind = catch
+-------------------------------
+C ⊢ rethrow l : [t1*] -> [t2*]
+
+
+bt = [t1*] -> [t2*]
+C, label {result [t2*], kind try} ⊢ instr* : [t1*] -> [t2*]
+(C_exn(x_i) = [t'_i*] -> [])^n
+(C, label { result [t2*], kind catch } ⊢ instr_i* : [t'_i*] -> [t2*])^n
+(C, label { result [t2*], kind catch } ⊢ instr'* : [] -> [t2*])^k
+(k=0 and n>0) or (k=1 and n≥0)
+------------------------------------------------------------------
+C ⊢ try bt instr* (catch x_i instr_i*)^n (catch_all instr'*)^k end : bt
+
+
+bt = [t1*] -> [t2*]
+C, label {result [t2*], kind try} ⊢ instr* : [t1*] -> [t2*]
+C_label(l) =/= ε
+C_label(l).kind = try
+------------------------------------------------------------
+C ⊢ try bt instr* delegate l : bt
+
+
+bt = [t1*] -> [t2*]
+C, label {result [t2*], kind try} ⊢ instr_1* : [t1*] -> [t2*]
+C, label [t2*] ⊢ instr_2* : [] -> []
+--------------------------------------------------------------
+C ⊢ try bt instr_1* unwind instr_2* end : bt
+```
+
+## Execution (Reduction)
+
+### Runtime structure
+
+Stores
+
+```
+S ::= {..., exn exninst*}
+```
+
+Exception Instances
+
+```
+exninst ::= {type exntype}
+```
+
+Module Instances
+
+```
+m ::= {..., (exn a)*}
+```
+
+Administrative Instructions
+
+```
+instr ::= ... | throw a | catch_n{a instr*}*{instr*}? instr* end
+        | delegate{l} instr* end | caught_m{a val^n} instr* end
+```
+
+Throw Contexts
+
+```
+T ::= val* [_] instr* | label_n{instr*} T end | caught_m{a val^n} T end | frame_n{F} T end
+```
+
+### Instructions
+
+
+```
+F; throw x  ↪  F; throw a  (iff F_exn(x) = a)
+
+caught_m{a val^n} B^{l+1}[rethrow l] end
+  ↪ caught_m{a val^n} B^{l+1}[val^n (throw a)] end
```

Comment:
To be honest this has been bothering me lately: as it is now, even in the main spec, any instructions could be between two labels in a `B^l`, or am I misunderstanding something?

Unless you mean that `rethrow`'s immediate `l` should only count the catches, which is [not the case](https://github.com/WebAssembly/exception-handling/issues/146#issuecomment-777349038) anymore (I missed this at first as it's a bit hidden in a different issue).

Reply:
I used `B^{l+1}` because `B^0` does not contain / is not pointing to- a label, while `rethrow 0` should point to a surrounding label.

Reply:
> To be honest this has been bothering me lately: as it is now, even in the main spec, any instructions could be between two labels in a `B^l`, or am I misunderstanding something?

Not sure I follow. A branch can skip over any instruction. What would be an example of an odd context?

> I used `B^{l+1}` because `B^0` does not contain / is not pointing to- a label, while `rethrow 0` should point to a surrounding label.

Intuitively, the `B^l` captures the l label binders that you _skip over_ with a branch. So if l=0 then this context should indeed not contain any label. The target label is the one explicit in the rule, surrounding B.

However:

> Also note that, in the execution rules, `caught_m` gets introduced with a label after it (the +1), to compensate for the evaporating try-label when catching the exception by a handler. Does this make sense?

Ah, yes, this explains why we need a +1 here: the catch actually is _outside_ its associated label, so we actually need to skip over that as well to get to the catch.

It's a bit odd, though. I suspect that this use of labels creates an issue with the soundness proof. In particular, what is the administrative typing rule for catch_n? It can't bind the label, since that is bound by label_n. But label_n won't attach the catch kind, so any contained rethrow won't type-check anymore after reducing try to catch, and we lose preservation.

Maybe this can be hacked by changing the reduction rule for try-catch so that it binds the label _outside_ the catch. Then the typing rule for catch_n could _amend_ the label kind of the innermost label in its environment. (And we wouldn't have the +1.)

Something to think about. I'd encourage you to write out the typing rules for the administrative instructions, to check that it all hangs together.

(I guess the more baroque nature of the new design is showing here.)